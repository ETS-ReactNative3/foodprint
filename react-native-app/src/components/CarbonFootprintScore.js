import { heightPercentageToDP as percentageHeight, widthPercentageToDP as percentageWidth } from 'react-native-responsive-screen';
import { VictoryPie } from 'victory-native';
import React from 'react';
import { ActivityIndicator, Image, StyleSheet, Text, View } from 'react-native';
import { Tooltip } from 'react-native-elements';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';

const CarbonFootprintScore = ({ loading, error, historyReport }) => {

  const getColorFromCarbonFootprint = (carbonFootprint) => {
    if (carbonFootprint < 4) {
      return 'forestgreen';
    } else if (carbonFootprint < 12) {
      return 'yellowgreen';
    } else if (carbonFootprint < 16) {
      return 'yellow';
    } else if (carbonFootprint < 20) {
      return 'orange';
    }
    return 'red';
  };

  const getSmileyFromCarbonFootprint = (carbonFootprint) => {
    if (carbonFootprint < 4) {
      return require('../images/sparkling-eyes-smiley.png');
    } else if (carbonFootprint < 12) {
      return require('../images/happy-smiley.png');
    } else if (carbonFootprint < 16) {
      return require('../images/shy-smiley.png');
    } else if (carbonFootprint < 20) {
      return require('../images/unimpressed-smiley.png');
    }
    return require('../images/crying-smiley.png');
  };

  const TOOLTIP_MESSAGE = 'This score corresponds to the average carbon footprint of ' +
    'all the items you have added to your history since you have started using Foodprint.';

  const renderTooltip = () => (
    <Tooltip popover={<Text style={styles.tooltipContent}>{TOOLTIP_MESSAGE}</Text>}
      backgroundColor={'green'}
      height={percentageHeight('20%')}
      width={percentageWidth('65%')}>
      <MaterialCommunityIcons name="help-circle" color={'grey'} size={percentageWidth('4%')} />
    </Tooltip>
  );

  return (
    <View>
      {loading || error || !historyReport?.userAvg ? (
        <View style={styles.loadingContainer}>
          <ActivityIndicator />
        </View>
      ) : (
          <View style={styles.graphContainer}>
            <Image
              source={getSmileyFromCarbonFootprint(historyReport.userAvg)}
              style={styles.image}
            />
            <View style={styles.scoreContainer}>
              <Text style={styles.score}>{Math.round(historyReport.userAvg)} units</Text>
              {renderTooltip()}
            </View>
            <VictoryPie
              data={[{ x: ' ', y: historyReport.userAvg }, { x: ' ', y: 26.7 - historyReport.userAvg }]}
              standalone={true}
              colorScale={[getColorFromCarbonFootprint(historyReport.userAvg), 'transparent']}
              startAngle={-90}
              endAngle={90}
              innerRadius={percentageHeight('16%')}
              height={percentageHeight('40%')}
            />
          </View>
        )}
    </View>
  );
};

const styles = StyleSheet.create({
  loadingContainer: { height: percentageHeight('29%'), alignItems: 'center', justifyContent: 'center' },
  graphContainer: { height: percentageHeight('29%'), alignItems: 'center' },
  tooltipContent: { color: 'white', fontSize: percentageWidth('4%') },
  image: { height: percentageHeight('10%'), width: percentageWidth('20%'), position: 'absolute', alignSelf: 'center', marginTop: percentageHeight('12%') },
  scoreContainer: { position: 'absolute', alignSelf: 'center', marginTop: percentageHeight('24%'), flexDirection: 'row' },
  score: { fontSize: percentageWidth('6%'), color: 'grey', margin: percentageWidth('1%') },
});

export default CarbonFootprintScore;
