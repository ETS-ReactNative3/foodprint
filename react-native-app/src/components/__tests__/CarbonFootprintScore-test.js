import React from 'react';
import Renderer from 'react-test-renderer';
import CarbonFootprintScore from '../CarbonFootprintScore';


const mockedHistoryReport = {
  userAvg: 17.25,
};

test('matches snapshot', async () => {
  const result = Renderer.create(<CarbonFootprintScore historyReport={mockedHistoryReport} />);

  expect(result.toJSON()).toMatchSnapshot();
});
