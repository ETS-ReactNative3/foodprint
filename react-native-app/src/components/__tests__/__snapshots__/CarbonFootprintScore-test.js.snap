// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 387,
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../src/images/unimpressed-smiley.png",
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "height": 133.5,
          "marginTop": 160,
          "position": "absolute",
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginTop": 320,
          "position": "absolute",
          "zIndex": 100,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "grey",
            "fontSize": 24,
          }
        }
      >
        17.3
         
        foodprints
      </Text>
      <View
        collapsable={false}
      >
        <TouchableOpacity
          activeOpacity={1}
          onPress={[Function]}
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "grey",
                  "fontSize": 24,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </TouchableOpacity>
        <Modal
          animationType="fade"
          hardwareAccelerated={false}
          onDismiss={[Function]}
          onRequestClose={[Function]}
          onShow={[Function]}
          transparent={true}
          visible={false}
        >
          <TouchableOpacity
            activeOpacity={1}
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "rgba(250, 250, 250, 0.70)",
                "flex": 1,
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "height": 0,
                    "left": 0,
                    "overflow": "visible",
                    "position": "absolute",
                    "top": 0,
                    "width": 0,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "grey",
                        "fontSize": 24,
                      },
                      undefined,
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  
                </Text>
              </View>
              <View
                style={
                  Object {
                    "left": -7.5,
                    "position": "absolute",
                    "top": -2,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "borderBottomColor": "#008000",
                      "borderBottomWidth": 15,
                      "borderLeftColor": "transparent",
                      "borderLeftWidth": 8,
                      "borderRightColor": "transparent",
                      "borderRightWidth": 8,
                      "borderStyle": "solid",
                      "height": 0,
                      "width": 0,
                    }
                  }
                />
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#008000",
                    "borderRadius": 10,
                    "display": "flex",
                    "flex": 1,
                    "height": 533.5,
                    "justifyContent": "center",
                    "left": -17.999,
                    "padding": 10,
                    "position": "absolute",
                    "top": 10,
                    "width": 487.5,
                  }
                }
                testID="tooltipPopoverContainer"
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 16,
                    }
                  }
                >
                  Foodprints is a unit we invented, but it is equivalent to kg of CO₂/kg of food, so the more unsustainable foods have a higher number.

It goes down when your diet is more sustainable, and doesn't take into quantities.
                </Text>
              </View>
            </View>
          </TouchableOpacity>
        </Modal>
      </View>
    </View>
    <View
      className="VictoryContainer"
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      pointerEvents="box-none"
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={533.5}
        bbWidth={750}
        focusable={false}
        height={533.5}
        meetOrSlice={0}
        minX={0}
        minY={0}
        role="img"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "height": "100%",
              "width": "100%",
            },
            Object {
              "opacity": 1,
            },
            Object {
              "flex": 0,
              "height": 533,
              "width": 750,
            },
          ]
        }
        vbHeight={533.5}
        vbWidth={750}
        width={750}
      >
        <RNSVGGroup
          fill={4278190080}
          fillOpacity={1}
          fillRule={1}
          font={Object {}}
          matrix={
            Array [
              1,
              0,
              0,
              1,
              0,
              0,
            ]
          }
          opacity={1}
          originX={0}
          originY={0}
          propList={Array []}
          rotation={0}
          scaleX={1}
          scaleY={1}
          skewX={0}
          skewY={0}
          stroke={null}
          strokeDasharray={null}
          strokeDashoffset={null}
          strokeLinecap={0}
          strokeLinejoin={0}
          strokeMiterlimit={4}
          strokeOpacity={1}
          strokeWidth={1}
          vectorEffect={0}
          x={0}
          y={0}
        >
          <RNSVGGroup
            fill={4278190080}
            fillOpacity={1}
            fillRule={1}
            font={Object {}}
            matrix={
              Array [
                1,
                0,
                0,
                1,
                0,
                0,
              ]
            }
            opacity={1}
            originX={0}
            originY={0}
            propList={Array []}
            rotation={0}
            scaleX={1}
            scaleY={1}
            skewX={0}
            skewY={0}
            stroke={null}
            strokeDasharray={null}
            strokeDashoffset={null}
            strokeLinecap={0}
            strokeLinejoin={0}
            strokeMiterlimit={4}
            strokeOpacity={1}
            strokeWidth={1}
            vectorEffect={0}
            x={0}
            y={0}
          >
            <RNSVGPath
              d="M-216.75,-2.654421937151888e-14A216.75,216.75,0,0,1,96.00900570143422,-194.3266150691201L94.56942430106669,-191.4128365271379A213.5,213.5,0,0,0,-213.5,-2.6146209161795992e-14Z"
              fill={4294944000}
              fillOpacity={1}
              fillRule={1}
              matrix={
                Array [
                  1,
                  0,
                  0,
                  1,
                  375,
                  266.75,
                ]
              }
              opacity={1}
              originX={0}
              originY={0}
              propList={
                Array [
                  "fill",
                ]
              }
              rotation={0}
              scaleX={1}
              scaleY={1}
              skewX={0}
              skewY={0}
              stroke={null}
              strokeDasharray={null}
              strokeDashoffset={null}
              strokeLinecap={0}
              strokeLinejoin={0}
              strokeMiterlimit={4}
              strokeOpacity={1}
              strokeWidth={1}
              vectorEffect={0}
              x={0}
              y={0}
            />
            <RNSVGPath
              d="M96.00900570143422,-194.3266150691201A216.75,216.75,0,0,1,216.75,-4.8128168117500536e-14L213.5,-4.7406523151494184e-14A213.5,213.5,0,0,0,94.56942430106669,-191.4128365271379Z"
              fill={0}
              fillOpacity={1}
              fillRule={1}
              matrix={
                Array [
                  1,
                  0,
                  0,
                  1,
                  375,
                  266.75,
                ]
              }
              opacity={1}
              originX={0}
              originY={0}
              propList={
                Array [
                  "fill",
                ]
              }
              rotation={0}
              scaleX={1}
              scaleY={1}
              skewX={0}
              skewY={0}
              stroke={null}
              strokeDasharray={null}
              strokeDashoffset={null}
              strokeLinecap={0}
              strokeLinejoin={0}
              strokeMiterlimit={4}
              strokeOpacity={1}
              strokeWidth={1}
              vectorEffect={0}
              x={0}
              y={0}
            />
            <RNSVGText
              content={null}
              dx={
                Array [
                  0,
                ]
              }
              dy={
                Array [
                  -2.0300000000000002,
                ]
              }
              fill={4278190080}
              fillOpacity={1}
              fillRule={1}
              font={Object {}}
              matrix={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
              name="pie-labels-0"
              opacity={1}
              originX={0}
              originY={0}
              propList={Array []}
              rotate={Array []}
              rotation={0}
              scaleX={1}
              scaleY={1}
              skewX={0}
              skewY={0}
              stroke={null}
              strokeDasharray={null}
              strokeDashoffset={null}
              strokeLinecap={0}
              strokeLinejoin={0}
              strokeMiterlimit={4}
              strokeOpacity={1}
              strokeWidth={1}
              vectorEffect={0}
              x={
                Array [
                  250,
                ]
              }
              y={
                Array [
                  65.75,
                ]
              }
            >
              <RNSVGTSpan
                content=" "
                dx={
                  Array [
                    0,
                  ]
                }
                dy={Array []}
                fill={4280624421}
                fillOpacity={1}
                fillRule={1}
                font={
                  Object {
                    "fontFamily": "Gill Sans",
                    "fontSize": 14,
                    "letterSpacing": "normal",
                    "textAnchor": "middle",
                  }
                }
                matrix={
                  Array [
                    1,
                    0,
                    0,
                    1,
                    0,
                    0,
                  ]
                }
                opacity={1}
                originX={0}
                originY={0}
                propList={
                  Array [
                    "fill",
                  ]
                }
                rotate={Array []}
                rotation={0}
                scaleX={1}
                scaleY={1}
                skewX={0}
                skewY={0}
                stroke={null}
                strokeDasharray={null}
                strokeDashoffset={null}
                strokeLinecap={0}
                strokeLinejoin={0}
                strokeMiterlimit={4}
                strokeOpacity={1}
                strokeWidth={1}
                vectorEffect={0}
                x={
                  Array [
                    250,
                  ]
                }
                y={Array []}
              />
            </RNSVGText>
            <RNSVGText
              content={null}
              dx={
                Array [
                  0,
                ]
              }
              dy={
                Array [
                  4.97,
                ]
              }
              fill={4278190080}
              fillOpacity={1}
              fillRule={1}
              font={Object {}}
              matrix={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
              name="pie-labels-1"
              opacity={1}
              originX={0}
              originY={0}
              propList={Array []}
              rotate={Array []}
              rotation={0}
              scaleX={1}
              scaleY={1}
              skewX={0}
              skewY={0}
              stroke={null}
              strokeDasharray={null}
              strokeDashoffset={null}
              strokeLinecap={0}
              strokeLinejoin={0}
              strokeMiterlimit={4}
              strokeOpacity={1}
              strokeWidth={1}
              vectorEffect={0}
              x={
                Array [
                  576,
                ]
              }
              y={
                Array [
                  141.75,
                ]
              }
            >
              <RNSVGTSpan
                content=" "
                dx={
                  Array [
                    0,
                  ]
                }
                dy={Array []}
                fill={4280624421}
                fillOpacity={1}
                fillRule={1}
                font={
                  Object {
                    "fontFamily": "Gill Sans",
                    "fontSize": 14,
                    "letterSpacing": "normal",
                    "textAnchor": "start",
                  }
                }
                matrix={
                  Array [
                    1,
                    0,
                    0,
                    1,
                    0,
                    0,
                  ]
                }
                opacity={1}
                originX={0}
                originY={0}
                propList={
                  Array [
                    "fill",
                  ]
                }
                rotate={Array []}
                rotation={0}
                scaleX={1}
                scaleY={1}
                skewX={0}
                skewY={0}
                stroke={null}
                strokeDasharray={null}
                strokeDashoffset={null}
                strokeLinecap={0}
                strokeLinejoin={0}
                strokeMiterlimit={4}
                strokeOpacity={1}
                strokeWidth={1}
                vectorEffect={0}
                x={
                  Array [
                    576,
                  ]
                }
                y={Array []}
              />
            </RNSVGText>
          </RNSVGGroup>
        </RNSVGGroup>
      </RNSVGSvgView>
      <View
        pointerEvents="box-none"
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "top": 0,
            "zIndex": 99,
          }
        }
      >
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={533.5}
          bbWidth={750}
          focusable={false}
          height={533.5}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              Object {
                "height": "100%",
                "overflow": "visible",
                "width": "100%",
              },
              Object {
                "opacity": 1,
              },
              Object {
                "flex": 0,
                "height": 533,
                "width": 750,
              },
            ]
          }
          vbHeight={533.5}
          vbWidth={750}
          width={750}
        >
          <RNSVGGroup
            fill={4278190080}
            fillOpacity={1}
            fillRule={1}
            font={Object {}}
            matrix={
              Array [
                1,
                0,
                0,
                1,
                0,
                0,
              ]
            }
            opacity={1}
            originX={0}
            originY={0}
            propList={Array []}
            rotation={0}
            scaleX={1}
            scaleY={1}
            skewX={0}
            skewY={0}
            stroke={null}
            strokeDasharray={null}
            strokeDashoffset={null}
            strokeLinecap={0}
            strokeLinejoin={0}
            strokeMiterlimit={4}
            strokeOpacity={1}
            strokeWidth={1}
            vectorEffect={0}
            x={0}
            y={0}
          />
        </RNSVGSvgView>
      </View>
    </View>
  </View>
</View>
`;
